core/shared/src/main/scala/zio/ZLayer.scala

```diff
@@ -2350,91 +2339,73 @@ object ZLayer extends ZLayerCompanionVersionSpecific {
+                        b
                       }
-
-                    ZIO.succeed((cached, map))
-                  case None =>
-                    for {
-                      observers    <- Ref.make(0)
-                      promise      <- Promise.make[E, (FiberRefs.Patch, ZEnvironment[B])]
-                      finalizerRef <- Ref.make[Exit[Any, Any] => UIO[Any]](_ => ZIO.unit)
-
-                      resource = ZIO.uninterruptibleMask { restore =>
-                                   for {
-                                     a          <- ZIO.environment[A]
-                                     outerScope  = scope
-                                     innerScope <- Scope.make
-                                     tp <-
-                                       restore(
-                                         layer
-                                           .scope(innerScope)
-                                           .flatMap(_.apply(self).diffFiberRefs)
-                                       ).exit.flatMap {
-                                         case e @ Exit.Failure(cause) =>
-                                           promise.failCause(cause) *> innerScope.close(e) *> ZIO
-                                             .failCause(cause)
-
-                                         case Exit.Success((patch, b)) =>
-                                           for {
-                                             _ <- finalizerRef.set { (e: Exit[Any, Any]) =>
-                                                    ZIO.whenZIO(observers.modify(n => (n == 1, n - 1)))(
-                                                      innerScope.close(e)
-                                                    )
-                                                  }
-                                             _ <- observers.update(_ + 1)
-                                             outerFinalizer <-
-                                               outerScope.addFinalizerExit(e => finalizerRef.get.flatMap(_.apply(e)))
-                                             _ <- promise.succeed((patch, b))
-                                           } yield b
-                                       }
-                                   } yield tp
-                                 }
-
-                      memoized = (
-                                   promise.await.onExit {
-                                     case Exit.Failure(_) => ZIO.unit
-                                     case Exit.Success(_) => observers.update(_ + 1)
-                                   },
-                                   (exit: Exit[Any, Any]) => finalizerRef.get.flatMap(_(exit))
-                                 )
-                    } yield (resource, if (layer.isFresh) map else map.updated(layer, memoized))
-
+                  case e @ Exit.Failure(cause) =>
+                    promise.failCause(cause) *> innerScope.close(e) *> ZIO.failCause(cause)
```

Comment:
You can avoid re-allocating the `Exit`.
```suggestion
                    promise.done(e) *> innerScope.close(e) *> ZIO.failCause(cause)
```

Won't the cause already be enriched with a stacktrace? Do we need `ZIO.failCause` here or can it just be `*> e`?

Reply:
I was thinking of changing it to `*> e` but was a bit worried to do so. Looking at it again I think it's quite safe, I'll change it

Reply:
Turns out we actually need it, a test is failing because of the `*> e` change so I reverted it to `ZIO.failCause(cause)`

Reply:
This is the assert that's failing:

https://github.com/zio/zio/blob/series/2.x/test-tests/shared/src/test/scala/zio/test/ZIOSpecAbstractSpec.scala#L44-L47

Enriching traces will be expensive in apps where layers are used at runtime for instantiation... obviously defer to better DX, but I don't know how much of a difference it really is, given the important failure comes from the layer itself?

Reply:
If a layer is reused in different places, the location where the layer failure happened will be relevant though